### ISSUE TITLE
AsyncIO can cause deadlocks in the InteractiveShell

### ISSUE THREAD

**Issue by sonthonaxrk** [ipython/ipython#12786]
I was looking for a way to delay module imports until some AsyncIO code has run.

Before anyone calls me mad, the reason I want to do this: is to check for an `ipykernel.comm.Comm` to have setup. I am waiting on an RPC call that verifies my Jupyter extension has loaded; this solves the problem that Jupyter extensions load asynchronously. When JavaScript comm target it registered it sends a success message back.

```
comm = Comm('my_ext')

@comm.on_msg
def on_message(msg):
     logger.debug('here')
```

```
def create_future():
    future = asyncio.Future()
    comm = Comm('my_ext')

    @comm.on_msg
    def on_message(msg):
        future.set_result(comm)
        comm.on_msg(None)

    return future

async def wait_for_future():
    return await create_future()

# Something along the lines of this...
loop.run_until_complete(wait_for_future())
```

In the context of a Jupyter Notebook: I wanted to have the above code block my module import (with a short timeout, obviously) until `on_message` was invoked.  I tried wrapping that code in an `asyncio.Future`, but ran into problems actually executing it because Jupyter has an event loop already running.

I tried to work around this with threads and `loop.call threadsafe`; then waiting for the thread to complete.  I also tried something similar with a nested event loop (async_nested).

### What I think the core problem is

The problem is that I can't wait for the `asyncio.Future` without synchronously blocking the code inside `run_cell_async`, pausing all other execution on the event loop. The deadlock can not be released because it depends on the `on_message` future resolving, but `Comm` messages are sent by that same event loop.

**Comment by sonthonaxrk** [comment]
Just to add I don't think this is an entirely insane use case.

Quite a few libraries do things like this, where they do a bit of IO on importation. Normally it is file IO. But it's conceivable that someone would need to do a little bit of AsyncIO networking.

Also, this code will only ever run in a Jupyter context. And it's meant to be userfriendly (non programmers), rather than strictly ideomatic.

**Comment by Carreau** [comment]
Yeah, the core of IPython is not completely async aware; and we do pretty big hack to even have a sort of async-exec. 

I'm not too sure how to fix all of that; it may even requires some change in the eval model of Python (and Jupyter), so that async execution can happen in the back. You may be able to solve that with thread though I rarely touch async these days so it might be tough.

**Comment by sonthonaxrk** [comment]
I'm certainly interested in solving it. I was researching ways to build GUI's within Jupyter, and the wacky async integration has been a real stumbling block.

I noticed another way it interferes with a slighly less insane usecase.

I wanted to be able to wait on GUI input, like this.

```python
import asyncio
import ipywidgets as widgets
from IPython.display import display


def wait_for_click(button):
    future = asyncio.Future()

    def getvalue(button):
        # make the new value available
        future.set_result(True)
        button.on_click(getvalue, remove=True)

    button.on_click(getvalue)
    return future

button = widgets.Button(description='Hello world')
click_future = wait_for_click()
```

In Jupyter then:


```
In  [1]: from my_module import *
           display(button)

Out [1]: ... button display ...


In  [2]: await click_future


However, this will never resolve if I run cell two.

It's really annoying, because I wanted to be able to start little GUI REPLs for workflow automation that required some human input.  Like a loop that gets items from an API, presents a GUI in the cell output, waits for input, and continues.

**Comment by sonthonaxrk** [comment]
@Carreau I've spent a bit of time on this, and I think I understand the problem. It's in the ipykernel.

https://github.com/ipython/ipykernel/blob/master/ipykernel/kernelbase.py#L360-L376

```python
    @gen.coroutine
    def dispatch_queue(self):
        """Coroutine to preserve order of message handling
        Ensures that only one message is processing at a time,
        even when the handler is async
        """

        while True:
            # ensure control stream is flushed before processing shell messages
            if self.control_stream:
                self.control_stream.flush()
            # receive the next message and handle it
            try:
                yield self.process_one()
            except Exception:
                self.log.exception("Error in message handler")
```

So, `process_one` effectively makes the IPython notebook single threaded (in terms of incoming ZeroMQ events).

Maybe there could be a seperate queue for comm events? Everything is already quite well done here, and that could be a simple solution :)

**Comment by sonthonaxrk** [comment]
https://github.com/ipython/ipykernel/pull/323

This is the original PR that introduced this work. I think it's all great amazing work. Wondering if you'd be open to managing comm events a bit differently :)

**Comment by Carreau** [comment]
if you touching comms you want to get feedback from the QuantStack folks and the ipywidget crowd; @SylvainCorlay, might be able point you to the right folks, and Xeus-Python might work for that kind of things ?

**Comment by sonthonaxrk** [comment]
@Carreau  I feel pretty confident about making the change. I feel that the asyncio code is still a bit immature (hasn't changed since it was first implemented).  

As I said in the Gitter channel: I think we should treat `comm_msg` as a special case that's always handled even if there's another cell running. While the other message types are definitely synchronous, like `execute_request`, Comm events are not. For example, cell execution obviously needs to happen atomically; but should a button click really be blocked by another cell doing an `await asyncio.sleep(10)`?

I tried out Xeus-Python to see if that handled things different. It doesn't, it also has a mutex on the io_pub queue.


https://github.com/jupyter-xeus/xeus-python/blob/bccb73190768bb435c30d406491458ccc9de15a6/test/xeus_client.cpp#L244-L250

```cpp
nl::json xeus_logger_client::pop_iopub_message()
{
    std::lock_guard<std::mutex> guard(m_queue_mutex);
    nl::json res = m_message_queue.back();
    m_message_queue.pop();
    return res;
}
```

**Comment by sonthonaxrk** [comment]
Xeus has the same issue, and I'm just scanning the source to see what the problem there is.

**Comment by sonthonaxrk** [comment]
@Carreau what's the process for these kinds of changes? If it's a feature, does it need to go through a committee and roadmap? I see Xeus is the reference implementation, does that mean changes have to go there first?

**Comment by sonthonaxrk** [comment]
@Carreau I've found solution.

I've written a Kernel demonstrating it. Now I can await comm events in my notebooks! It's just a little hack.

https://gist.github.com/sonthonaxrk/ecf2222a3e8f0035928467c36c443428

```
    def schedule_dispatch(self, priority, dispatch, *args):
        """
        Changes the schedule_dispatch dispatch method to
        always dispatch comm events.
        """
        # quick and dirty checks for message type

        if len(args) == 2:  # SHELL_PRIORITY
            stream, unparsed_msg = args
            indent, msg = self._parse_message(unparsed_msg)
            new_args = (stream, msg, indent)
        elif len(args) == 1: # CONTROL_PRIORITY
            # One arg
            (unparsed_msg,) = args
            indent, msg = self._parse_message(unparsed_msg)
            new_args = (msg, indent)
        elif len(args) == 0: # ABORT_PRIORITY
            new_args = args

        if new_args and msg['header']['msg_type'] in self.comm_msg_types:
            return dispatch(*new_args)
        else:
            idx = next(self._message_counter)

            self.msg_queue.put_nowait(
                (
                    priority,
                    idx,
                    dispatch,
                    new_args,
                )
            )
            # ensure the eventloop wakes up
            self.io_loop.add_callback(lambda: None)

```

----

However, I'm thinking about a more robust way of doing this.

**Comment by sonthonaxrk** [comment]
@Carreau https://github.com/sonthonaxrk/async_gui_ipython_kernel

I've written a kernel subclass that kinda fixes this issue.
